### PR TITLE
tests: add unit tests for cmd_whoami

### PR DESCRIPTION
**Pull request by ardaguclu**
Added unit tests for cmd_whoami. It would be great adding tests for logged users email address verification either. But logging in cmd_whoami might be out of responsibility of it. If there is builtin mocked logging mechanism, it will be great.

### CODE REVIEW

**Comment by ardaguclu** [comment]
Thanks for your feedback and help. I made additional changes according to your comments.

File: cmd/snap/cmd_whoami_test.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
+package main_test
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]: The copyright header is missing here.

File: cmd/snap/cmd_whoami_test.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+func (s *SnapSuite) TestWhoamiEmptyAuthFile(c *C) {
+	s.Login(c)
+
+	wErr := osutil.AtomicWriteFile(s.AuthFile, []byte(``), 0600, 0)
+	c.Assert(wErr, IsNil)
+
+	_, err := snap.Parser(snap.Client()).ParseArgs([]string{"whoami"})
+	c.Assert(err, NotNil)
+	c.Check(err.Error(), Equals, "EOF")
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]: Just a style suggestion, we rarely call `.Error`, on error objects. Instead we almost always use `ErrorMatches` for that.

```suggestion
	c.Check(err, ErrorMatches, "EOF")
```

File: cmd/snap/cmd_whoami_test.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+func (s *SnapSuite) TestWhoamiEmptyAuthFile(c *C) {
+	s.Login(c)
+
+	wErr := osutil.AtomicWriteFile(s.AuthFile, []byte(``), 0600, 0)
+	c.Assert(wErr, IsNil)
+
+	_, err := snap.Parser(snap.Client()).ParseArgs([]string{"whoami"})
+	c.Assert(err, NotNil)
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]: You can drop the non-nil assertion check if you use `ErrorMatches`

File: cmd/snap/cmd_whoami_test.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+	_, err := snap.Parser(snap.Client()).ParseArgs([]string{"whoami"})
+	c.Assert(err, IsNil)
+	c.Check(s.Stdout(), Equals, "email: -\n")
+}
+
+func (s *SnapSuite) TestWhoamiExtraParamError(c *C) {
+	_, err := snap.Parser(snap.Client()).ParseArgs([]string{"whoami", "test"})
+	c.Assert(err, NotNil)
+	c.Check(err.Error(), Equals, "too many arguments for command")
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]:
```suggestion
	c.Check(err, ErrorMatches, "too many arguments for command")
```

File: cmd/snap/cmd_whoami_test.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	_, err := snap.Parser(snap.Client()).ParseArgs([]string{"whoami"})
+	c.Assert(err, IsNil)
+	c.Check(s.Stdout(), Equals, "email: hello@mail.com\n")
+
+	s.Logout(c)
+}
+
+func (s *SnapSuite) TestWhoamiNotLoggedInUser(c *C) {
+	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
+		c.Errorf("whoami does not hit the API, it just reads ~/.snap/auth.json (which is mocked in the tests)")
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]: I think it's fine to just say `panic("unexpected call to snapd API")` here

File: cmd/snap/cmd_whoami_test.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	})
+
+	s.Login(c)
+
+	_, err := snap.Parser(snap.Client()).ParseArgs([]string{"whoami"})
+	c.Assert(err, IsNil)
+	c.Check(s.Stdout(), Equals, "email: hello@mail.com\n")
+
+	s.Logout(c)
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]: It is unclear what this logout is for. If it is to clean up some state it should be paired with `s.Login` call above in the style of `defer s.Logout(c)`

File: cmd/snap/cmd_whoami_test.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/snapcore/snapd/osutil"
+
+	. "gopkg.in/check.v1"
+
+	snap "github.com/snapcore/snapd/cmd/snap"
+)
+
+func (s *SnapSuite) TestWhoamiLoggedInUser(c *C) {
+	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
+		c.Errorf("whoami does not hit the API, it just reads ~/.snap/auth.json (which is mocked in the tests)")
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]: Same remark about the error in API server as below.

File: cmd/snap/cmd_whoami_test.go (hunk 2 above)
Comment (ardaguclu) [reply cmd/snap/cmd_whoami_test.go]: Changed.

File: cmd/snap/cmd_whoami_test.go (hunk 1 above)
Comment (ardaguclu) [reply cmd/snap/cmd_whoami_test.go]: Added.

File: cmd/snap/cmd_whoami_test.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]:
```suggestion
 * Copyright (C) 2019 Canonical Ltd
```

File: cmd/snap/cmd_whoami_test.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+func (s *SnapSuite) TestWhoamiExtraParamError(c *C) {
+	_, err := snap.Parser(snap.Client()).ParseArgs([]string{"whoami", "test"})
+	c.Check(err, ErrorMatches, "too many arguments for command")
+}
+
+func (s *SnapSuite) TestWhoamiEmptyAuthFile(c *C) {
+	s.Login(c)
+	defer s.Logout(c)
+
+	wErr := osutil.AtomicWriteFile(s.AuthFile, []byte(``), 0600, 0)
```
Comment (zyga) [inline cmd/snap/cmd_whoami_test.go]: You can reuse error variables. We have some checks that verify that an assignment is not _ineffectual_, that is, assigned but then never used. This allows you to write code like:
```
err := foo()
c.Assert(err, IsNil)
err = bar()
c.Assert(bar, ErrorMatches, `...`)
```
without having to invent variable names for each err.

File: cmd/snap/cmd_whoami_test.go (hunk 8 above)
Comment (ardaguclu) [reply cmd/snap/cmd_whoami_test.go]: Changed.

File: cmd/snap/cmd_whoami_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+func (s *SnapSuite) TestWhoamiEmptyAuthFile(c *C) {
+	s.Login(c)
+	defer s.Logout(c)
+
+	err := osutil.AtomicWriteFile(s.AuthFile, []byte(``), 0600, 0)
+	c.Assert(err, IsNil)
+
+	_, err = snap.Parser(snap.Client()).ParseArgs([]string{"whoami"})
+	c.Check(err, ErrorMatches, "EOF")
```
Comment (mvo5) [inline cmd/snap/cmd_whoami_test.go]: (nitpick) this would be io.EOF I think - but thats fine :)

**Pull request merged**